// gd-sprest/build/mapper/custom/graph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.graph = void 0;
var utils_1 = require("../../utils");
/**
 * Graph
 */
exports.graph = {
    // Me
    me: {
        requestType: utils_1.RequestType.GraphGet
    },
    // Group
    group: {
        argNames: ["id"],
        name: "groups/[[id]]",
        requestType: utils_1.RequestType.GraphGetReplace
    },
    // Groups
    groups: {
        name: "groups",
        requestType: utils_1.RequestType.GraphGet
    },
    // List
    list: {
        argNames: ["siteId", "id"],
        name: "sites/[[siteId]]/lists/[[id]]",
        requestType: utils_1.RequestType.GraphGetReplace
    },
    // Lists
    lists: {
        argNames: ["siteId"],
        name: "sites/[[siteId]]/lists",
        requestType: utils_1.RequestType.GraphGetReplace
    },
    // Queries the collection
    query: {
        argNames: ["oData"],
        requestType: utils_1.RequestType.OData
    },
    // Root Site
    root: {
        requestType: utils_1.RequestType.GraphGet
    },
    // Site
    site: {
        argNames: ["id"],
        name: "sites/[[id]]",
        requestType: utils_1.RequestType.GraphGetReplace
    },
    // Sites
    sites: {
        name: "sites",
        requestType: utils_1.RequestType.GraphGet
    },
    // User
    user: {
        argNames: ["id"],
        name: "users/[[id]]",
        requestType: utils_1.RequestType.GraphGetReplace
    },
    // Users
    users: {
        requestType: utils_1.RequestType.GraphGet
    },
};
